meta-lmp-bsp/dynamic-layers/freescale-layer/recipes-bsp/imx-atf/imx-atf/0001-plat-imx8m-obtain-boot-set-from-bootrom-even-log.patch

```diff
@@ -0,0 +1,96 @@
+From 58cd2feeb07e496c7884f6655f258eca2a25b7ff Mon Sep 17 00:00:00 2001
+From: Igor Opaniuk <igor.opaniuk@foundries.io>
+Date: Fri, 18 Mar 2022 14:52:22 -0300
+Subject: [PATCH] plat: imx8m: obtain boot set from bootrom even log
+
+Both i.MX8MP and i.MX8MN don't have a dedicated bit which indicates
+a boot image set booted by BootROMv According to AN12853 [1] i.MX ROMs Log
+Events, it is possible to determine whether fallback event occurred
+by parsing the ROM event log. In case ROM event ID 0x51 is present,
+fallback event did occur.
+
+The code is based on U-Boot implementation [2].
+
+[1]: https://www.nxp.com/webapp/Download?colCode=AN12853
+[2]: https://github.com/u-boot/u-boot/commit/a5ee05cf7180b411ffdf148ca8cb220c029f2e19
+Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>
+---
+ plat/imx/imx8m/gpc_common.c | 55 +++++++++++++++++++++++++++++++++++++
+ 1 file changed, 55 insertions(+)
+
+diff --git a/plat/imx/imx8m/gpc_common.c b/plat/imx/imx8m/gpc_common.c
+index 9e057b7df..e9e4e57c6 100644
+--- a/plat/imx/imx8m/gpc_common.c
++++ b/plat/imx/imx8m/gpc_common.c
+@@ -362,6 +362,55 @@ int imx_gpc_handler(uint32_t smc_fid, u_register_t x1, u_register_t x2, u_regist
+ #define MCU_RDC_MAGIC "mcu_rdc"
+ #endif
+ 
++static bool is_secondary_boot(void)
++{
++	uint32_t *rom_log_addr = (uint32_t *)0x9e0;
++	bool is_secondary = false;
++	uint32_t *rom_log;
++	uint8_t event_id;
++
++	/* If the ROM event log pointer is not valid. */
++	if (*rom_log_addr < 0x900000 || *rom_log_addr >= 0xb00000 ||
++	    *rom_log_addr & 0x3)
++		return false;
++
++	/* Parse the ROM event ID version 2 log */
++	rom_log = (uint32_t *)(uintptr_t)(*rom_log_addr);
++	for (size_t i = 0; i < 128; i++) {
++		event_id = rom_log[i] >> 24;
++		switch (event_id) {
++		case 0x00: /* End of list */
++			return is_secondary;
++		/* Log entries with 1 parameter, skip 1 */
++		case 0x80: /* Start to perform the device initialization */
++		case 0x81: /* The boot device initialization completes */
++		case 0x82: /* Start to execute boot device driver pre-config */
++		case 0x8f: /* The boot device initialization fails */
++		case 0x90: /* Start to read data from boot device */
++		case 0x91: /* Reading data from boot device completes */
++		case 0x9f: /* Reading data from boot device fails */
++			i += 1;
++			continue;
++		/* Log entries with 2 parameters, skip 2 */
++		case 0xa0: /* Image authentication result */
++		case 0xc0: /* Jump to the boot image soon */
++
++			i += 2;
++			continue;
++		/* Boot from the primary boot image */
++		case 0x50:
++			is_secondary = false;
++			continue;
++		/* Boot from the secondary boot image */
++		case 0x51:
++			is_secondary = true;
++			continue;
```

Comment:
only a question. do we expect that the event log would have multiple 0x50 and or 0x51 meaning could it show both?

Reply:
yes, when secondary image is loaded, BootROM reports 0x50 first, but then 0x51

Reply:
So we always rely on the latest value

Reply:
u-boot=> md 009112f8 # show events from event buffer
009112f8: e0000000 22000000 31000000 40000000  ......."...1...@
00911308: 61000000 82000000 00000ed9 83000000  ...a............
00911318: 80000000 00000ee2 81000000 00009c37  ............7...
00911328: **50000000** 90000000 00000000 91000000  ...P............
00911338: 00009ebc a1e00001 **51000000** 90000000  ...........Q....
00911348: 00000000 91000000 0000a155 90000000  ........U.......
00911358: 00000008 91000000 0000d357 a0000000  ........W.......
00911368: 00cf1133 0000d466 c0000000 00920000  3...f...........
00911378: 0000e758 00000000 00000000 00000000  X...............
00911388: 00000000 00000000 00000000 00000000  ................
00911398: 00000000 00000000 00000000 00000000  ................
009113a8: 00000000 00000000 00000000 00000000  ................
009113b8: 00000000 00000000 00000000 00000000  ................
009113c8: 00000000 00000000 00000000 00000000  ................
009113d8: 00000000 00000000 00000000 00000000  ................
009113e8: 00000000 00000000 00000000 00000000  ................